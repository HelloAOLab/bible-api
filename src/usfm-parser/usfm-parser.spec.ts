import { describe, it, expect, beforeEach } from 'vitest';


import { isDigit,
    isWhitespace,
    loc,
    marker,
    t,
    UsfmParser,
    usfmParser,
    UsfmTokenizer,
    whitespace,
    word
} from './usfm-parser';

import { readFile } from 'fs/promises';
import { resolve } from 'path';

describe('UsfmTokenizer', () => {
    let tokenizer: UsfmTokenizer;

    beforeEach(() => {
        tokenizer = new UsfmTokenizer();
    });

    describe('tokenize()', () => {
        describe('markers', () => {
            it('should be able to parse single character markers', () => {
                const tokens = tokenizer.tokenize(`\\a`);

                expect(tokens).toEqual([
                    t(loc(0, 2), 'marker'),
                ]);
            });

            it('should be able to parse multiple character markers', () => {
                const tokens = tokenizer.tokenize(`\\abc`);

                expect(tokens).toEqual([
                    t(loc(0, 4), 'marker'),
                ]);
            });

            it('should be able to parse markers with numbers', () => {
                const tokens = tokenizer.tokenize(`\\abc123`);

                expect(tokens).toEqual([
                    t(loc(0, 7), 'marker'),
                ]);
            });

            it('should be able to parse end markers', () => {
                const tokens = tokenizer.tokenize(`\\abc*`);

                expect(tokens).toEqual([
                    t(loc(0, 5), 'marker'),
                ]);
            });

            it('should be able to parse end markers with numbers', () => {
                const tokens = tokenizer.tokenize(`\\abc123*`);

                expect(tokens).toEqual([
                    t(loc(0, 8), 'marker'),
                ]);
            });

            it('should not throw an error if the marker does not have a command', () => {
                const tokens = tokenizer.tokenize(`\\`);

                expect(tokens).toEqual([
                    t(loc(0, 1), 'marker')
                ]);
            });

            it('should not throw an error if the marker has a number but not a command', () => {
                const tokens = tokenizer.tokenize(`\\1`);
                expect(tokens).toEqual([
                    t(loc(0, 2), 'marker')
                ]);
            });

            it('should be able to parse end markers that have text after them', () => {
                const tokens = tokenizer.tokenize(`\\abc*z`);

                expect(tokens).toEqual([
                    t(loc(0, 5), 'marker'),
                    t(loc(5, 6), 'word'),
                ]);
            });

            it('should be able to parse markers that have a space after them', () => {
                const tokens = tokenizer.tokenize(`\\abc z`);

                expect(tokens).toEqual([
                    t(loc(0, 4), 'marker'),
                    t(loc(4, 5), 'whitespace'),
                    t(loc(5, 6), 'word'),
                ]);
            });

            it('should be able to parse markers that have a tab after them', () => {
                const tokens = tokenizer.tokenize(`\\abc\tz`);

                expect(tokens).toEqual([
                    t(loc(0, 4), 'marker'),
                    t(loc(4, 5), 'whitespace'),
                    t(loc(5, 6), 'word'),
                ]);
            });

            it('should be able to parse markers that have a newline after them', () => {
                const tokens = tokenizer.tokenize(`\\abc\nz`);

                expect(tokens).toEqual([
                    t(loc(0, 4), 'marker'),
                    t(loc(4, 5), 'whitespace'),
                    t(loc(5, 6), 'word'),
                ]);
            });
        });

        describe('words', () => {
            it('should be able to parse characters into words', () => {
                const tokens = tokenizer.tokenize(`abc`);

                expect(tokens).toEqual([
                    t(loc(0, 3), 'word'),
                ]);
            });
        });

        describe('whitespace', () => {
            it('should be able to parse space into whitespace', () => {
                const tokens = tokenizer.tokenize(` `);

                expect(tokens).toEqual([
                    t(loc(0, 1), 'whitespace'),
                ]);
            });

            it('should be able to parse newlines into whitespace', () => {
                const tokens = tokenizer.tokenize(`\n`);

                expect(tokens).toEqual([
                    t(loc(0, 1), 'whitespace'),
                ]);
            });
            
            it('should be able to parse carriage returns into whitespace', () => {
                const tokens = tokenizer.tokenize(`\r`);

                expect(tokens).toEqual([
                    t(loc(0, 1), 'whitespace'),
                ]);
            });

            it('should be able to parse tabs into whitespace', () => {
                const tokens = tokenizer.tokenize(`\t`);

                expect(tokens).toEqual([
                    t(loc(0, 1), 'whitespace'),
                ]);
            });

            it('should be able to parse multiple whitespace into whitespace', () => {
                const tokens = tokenizer.tokenize(`\t \r\n`);

                expect(tokens).toEqual([
                    t(loc(0, 4), 'whitespace'),
                ]);
            });
        });
    });
});

describe('UsfmParser', () => {
    let parser: UsfmParser;

    beforeEach(() => {
        parser = new UsfmParser();
    });

    describe('tokenize()', () => {
        it('should be able to tokenize single character markers', () => {
            const tokens = parser.tokenize(`\\a`);

            expect(tokens).toEqual([
                marker(loc(0, 2), '\\a', null, 'start')
            ]);
        });

        it('should be able to tokenize multiple character markers', () => {
            const tokens = parser.tokenize(`\\abc`);

            expect(tokens).toEqual([
                marker(loc(0, 4), '\\abc', null, 'start')
            ]);
        });

        it('should be able to parse markers with numbers', () => {
            const tokens = parser.tokenize(`\\abc123`);

            expect(tokens).toEqual([
                marker(loc(0, 7), '\\abc', 123, 'start'),
            ]);
        });

        it('should be able to parse end markers', () => {
            const tokens = parser.tokenize(`\\abc*`);

            expect(tokens).toEqual([
                marker(loc(0, 5), '\\abc', null, 'end'),
            ]);
        });

        it('should be able to parse end markers with numbers', () => {
            const tokens = parser.tokenize(`\\abc123*`);

            expect(tokens).toEqual([
                marker(loc(0, 8), '\\abc', 123, 'end'),
            ]);
        });

        it('should throw an error if the marker does not have a command', () => {
            expect(() => {
                parser.tokenize(`\\`);
            }).toThrowError();
        });

        it('should throw an error if the marker has a number but not a command', () => {
            expect(() => {
                parser.tokenize(`\\1`);
            }).toThrowError();
        });

        it('should be able to parse end markers that have text after them', () => {
            const tokens = parser.tokenize(`\\abc*z`);

            expect(tokens).toEqual([
                marker(loc(0, 5), '\\abc', null, 'end'),
                word(loc(5, 6), 'z'),
            ]);
        });

        it('should be able to parse markers that have a space after them', () => {
            const tokens = parser.tokenize(`\\abc z`);

            expect(tokens).toEqual([
                marker(loc(0, 4), '\\abc'),
                whitespace(loc(4, 5), ' '),
                word(loc(5, 6), 'z'),
            ]);
        });

        it('should be able to parse markers that have a tab after them', () => {
            const tokens = parser.tokenize(`\\abc\tz`);

            expect(tokens).toEqual([
                marker(loc(0, 4), '\\abc'),
                whitespace(loc(4, 5), '\t'),
                word(loc(5, 6), 'z'),
            ]);
        });

        it('should be able to parse markers that have a newline after them', () => {
            const tokens = parser.tokenize(`\\abc\nz`);

            expect(tokens).toEqual([
                marker(loc(0, 4), '\\abc'),
                whitespace(loc(4, 5), '\n'),
                word(loc(5, 6), 'z'),
            ]);
        });
    });

    describe('parse()', () => {
        it('should produce a list of chapters', () => {
            const tree = parser.parse(`\\c 1
                \\v 1 In the beginning God created the heavens and the earth.
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            expect(tree).toEqual({
                type: 'root',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'In the beginning God created the heavens and the earth.'
                                ]
                            },
                            { 
                                type: 'verse', 
                                number: 2,
                                content: [
                                    'Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.'
                                ]
                            }
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'Thus the heavens and the earth were completed in all their vast array.'
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should infer the first verse of a chapter', () => {
            const tree = parser.parse(`\\c 1
                \\q1 Blessed is the man 
                \\q2 who does not walk in the counsel of the wicked, 
                \\q1 or set foot on the path of sinners, 
                \\q2 or sit in the seat of mockers. 
                \\q1
                \\v 2 But his delight is in the Law of the LORD, 
                \\q2 and on His law he meditates day and night. 
                \\q1
                \\c 2
                \\q1 Why do the nations rage
                \\q2 and the peoples plot in vain? 
                \\q1 
            `);

            expect(tree).toEqual({
                type: 'root',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    { text: 'Blessed is the man', poem: 1 },
                                    { text: 'who does not walk in the counsel of the wicked,', poem: 2 },
                                    { text: 'or set foot on the path of sinners,', poem: 1 },
                                    { text: 'or sit in the seat of mockers.', poem: 2 }
                                ]
                            },
                            { 
                                type: 'verse',
                                number: 2,
                                content: [
                                    { text: 'But his delight is in the Law of the LORD,', poem: 1 },
                                    { text: 'and on His law he meditates day and night.', poem: 2 }
                                ]
                            }
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    { text: 'Why do the nations rage', poem: 1 },
                                    { text: 'and the peoples plot in vain?', poem: 2 }
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should support line breaks', () => {
            const tree = parser.parse(`\\c 1
                \\v 1 In the beginning God created the heavens and the earth.
                \\b
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\b
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            expect(tree).toEqual({
                type: 'root',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'In the beginning God created the heavens and the earth.'
                                ]
                            },
                            {
                                type: 'line_break'
                            },
                            { 
                                type: 'verse', 
                                number: 2,
                                content: [
                                    'Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.'
                                ]
                            },
                            {
                                type: 'line_break'
                            },
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'Thus the heavens and the earth were completed in all their vast array.'
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should support ID tags', () => {
            const tree = parser.parse(`
                \\id GEN - Berean Study Bible
                \\c 1
                \\v 1 In the beginning God created the heavens and the earth.
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            expect(tree).toEqual({
                type: 'root',
                id: 'GEN',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'In the beginning God created the heavens and the earth.'
                                ]
                            },
                            { 
                                type: 'verse', 
                                number: 2,
                                content: [
                                    'Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.'
                                ]
                            }
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'Thus the heavens and the earth were completed in all their vast array.'
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should support major title headings', () => {
            const tree = parser.parse(`
                \\mt Genesis
                \\c 1
                \\v 1 In the beginning God created the heavens and the earth.
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            expect(tree).toEqual({
                type: 'root',
                title: 'Genesis',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'In the beginning God created the heavens and the earth.'
                                ]
                            },
                            { 
                                type: 'verse', 
                                number: 2,
                                content: [
                                    'Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.'
                                ]
                            }
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'Thus the heavens and the earth were completed in all their vast array.'
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should support section headings in chapters', () => {
            const tree = parser.parse(`\\c 1
                \\s1 The Creation
                \\v 1 In the beginning God created the heavens and the earth.
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            expect(tree).toEqual({
                type: 'root',
                content: [
                    {
                        type: 'chapter',
                        number: 1,
                        content: [
                            {
                                type: 'heading',
                                content: [
                                    'The Creation'
                                ]
                            },
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'In the beginning God created the heavens and the earth.'
                                ]
                            },
                            { 
                                type: 'verse', 
                                number: 2,
                                content: [
                                    'Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.'
                                ]
                            }
                        ]
                    },
                    {
                        type: 'chapter',
                        number: 2,
                        content: [
                            { 
                                type: 'verse', 
                                number: 1,
                                content: [
                                    'Thus the heavens and the earth were completed in all their vast array.'
                                ]
                            },
                        ]
                    }
                ]
            });
        });

        it('should support section headings outside of chapters', () => {
            const tree = parser.parse(`\\s1 The Creation
            `);

            expect(tree).toEqual({
                type: 'root',
                content: [
                    {
                        type: 'heading',
                        content: [
                            'The Creation'
                        ]
                    }
                ]
            });
        });
        
        describe('Bible', () => {
            describe('BSB', () => {
                const cases = [
                    ['bsb/01GENBSB.usfm', 50] as const
                ];

                it.each(cases)('should be able to parse %s', async (file, expectedChapters) => {
                    const filePath = resolve(__dirname, '..', '..', 'bible', file);
                    const data = await readFile(filePath, { encoding: 'utf-8' });
                    const parsed = parser.parse(data);
                    const numChapters = parsed.content.filter(c => c.type === 'chapter').length;
                    expect(numChapters).toBe(expectedChapters);
                });
            });
        });
    });

    describe('renderMarkdown()', () => {
        it('should create basic markdown from a parse tree', () => {
            const tree = parser.parse(`
                \\mt Genesis
                \\c 1
                \\v 1 In the beginning God created the heavens and the earth.
                \\v 2 Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters.
                \\c 2
                \\v 1 Thus the heavens and the earth were completed in all their vast array.
            `);

            const md = parser.renderMarkdown(tree);
            expect(md).toMatchSnapshot();
        });
    });
});

describe('isDigit()', () => {
    it('should retrun true for all the ascii digits', () => {
        for(let d = 0; d < 10; d++) {
            let digit = d.toString();

            expect(isDigit(digit)).toBe(true);
        }
    });

    it('should return true for all digit code points in a string', () => {
        let str = '0123456789';

        for (let d of str) {
            expect(isDigit(d)).toBe(true);
        }
    });

    it('should return false for non digits', () => {
        let str = 'abcdefghijkflmnopqrstuvwxyz';

        for (let c of str) {
            expect(isDigit(c)).toBe(false);
        }
    });

    it('should return false for strings longer than length 1', () => {
        let strings = [
            '0a',
            '0b',
            '9a',
            '9b',
        ];

        for (let str of strings) {
            expect(isDigit(str)).toBe(false);
        }
    });
});

describe('isWhitespace()', () => {
    it('should return true for common whitespace characters', () => {
        let chars = [' ', '\n', '\t', '\r'];
        for(let char of chars) {
            expect(isWhitespace(char)).toBe(true);
        }
    });

    it('should return false for non-whitespace characters', () => {
        let str = 'abcdefghijkflmnopqrstuvwxyz0123456789';

        for (let c of str) {
            expect(isWhitespace(c)).toBe(false);
        }
    });
});