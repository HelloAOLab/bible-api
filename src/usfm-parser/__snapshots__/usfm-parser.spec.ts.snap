// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsfmParser parse() Bible should consistently parse arbnav/02-GENarbnav.usfm 1`] = `"7a123ac0509f1a850032c97396f97525cae9c9606aac027b882ab3b6bc7f5f91"`;

exports[`UsfmParser parse() Bible should consistently parse arbnav/70-MATarbnav.usfm 1`] = `"0e686c0ca576605002b30d16f1ff76c9845e75d183fb60b362130fbcfb6aa273"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/01GENBSB.usfm 1`] = `"8fe2cb8bc2e60019f519ae34426edc7c1e99ab590dedd84997f01e888ae2f6c0"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/02EXOBSB.usfm 1`] = `"370e1c0af26668480816084ac95723275be0ebf64a3885dbc155bd0ada05c597"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/03LEVBSB.usfm 1`] = `"0ee5957d9f33fb0a80ab3724255d8af9a7d01bf49768299ef6970e0d6482f220"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/04NUMBSB.usfm 1`] = `"7b053622294aaad5fe617899be32c6a24cf8bac11f3bc1d13ebb026620301f25"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/05DEUBSB.usfm 1`] = `"01d240dec29f983218d4d8aa802ecf228a9ae2f39f9575a67b2bfb348e689ce7"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/06JOSBSB.usfm 1`] = `"1ea0734371accb1355d28dc8ff1f5212dbe3dd886daa1b91c21ea6465135e0c3"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/07JDGBSB.usfm 1`] = `"ea8c345f2d4fe2ba00df4aad280705cfc00f7090e14298b7d3205290f1524c2c"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/08RUTBSB.usfm 1`] = `"1a5d76fcdd1f2ad7fb0694d5fbc809bd9d632d84823d5037720e413df5d7c6d5"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/091SABSB.usfm 1`] = `"20da49b918c33014dd0cd2f9eceee5faba84df4a710fef8ca73378ff01509e91"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/15EZRBSB.usfm 1`] = `"496d7e9e790f7fb068b3845c32326c0a00c2d22732d19ecb633fe2bd12082b14"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/16NEHBSB.usfm 1`] = `"21144c057940ea21151e8783b67e285b1bb59464bd95e1ba34b0774c4d257ef3"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/17ESTBSB.usfm 1`] = `"f0237ba811d06ee1c3e91c4e72b71ed87720ec592d3e316f938585a5b2ea3a34"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/18JOBBSB.usfm 1`] = `"d60c018cf0caa297f0c14aeb49352fd829445a42e3c6c9b79054dba5fa6b3062"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/19PSABSB.usfm 1`] = `"c9daf723c4a444f0e8f1cffe2e015c9782d537f967a3cc3d595b0e831dea7110"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/41MATBSB.usfm 1`] = `"de1676ea1d79b47b0171fd6a3ca24f011fa65617cb97c33b023c7ae6bae8b8f7"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/102SABSB.usfm 1`] = `"d8b901b95aad00f56ab75c0ec98917bd7f281fcc0ef5a66d64610fab4ab19d8c"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/111KIBSB.usfm 1`] = `"d04490714c039beeb28c79dff55511ffb9c957a6fb140abb0809fadec3171824"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/122KIBSB.usfm 1`] = `"bc9995cd4114390ceb9e12c72c2a7619b059288c36dbcc6adab25428ff95a431"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/131CHBSB.usfm 1`] = `"44f7d2dbdc3394df31b501f999b41ad71a6825fea3a6f5dfc606de6234dae400"`;

exports[`UsfmParser parse() Bible should consistently parse bsb/142CHBSB.usfm 1`] = `"7775ddb8b5af29c0175b7962ea3297c16243a109960cda2c4602cf6d179a3aed"`;

exports[`UsfmParser parse() Bible should consistently parse engwebp/02-GENengwebp.usfm 1`] = `"dfaf30edb8379f692d9f277f57ad43da08e9199d8d51d9514131d866a171c725"`;

exports[`UsfmParser parse() Bible should consistently parse engwebp/03-EXOengwebp.usfm 1`] = `"11242ae1043f9cd8d43d41cdce4f354e9060f15459cbd1db9e60694899674fb8"`;

exports[`UsfmParser parse() Bible should consistently parse engwebp/70-MATengwebp.usfm 1`] = `"8a1d9749fd58fa43f9966833951d4b50b82915f896f786b8b5baaf43a115669b"`;

exports[`UsfmParser renderMarkdown() should create basic markdown from a parse tree 1`] = `
"# Genesis
### 1
<em>1</em>In the beginning God created the heavens and the earth. 
<em>2</em>Now the earth was formless and void, and darkness was over the surface of the deep. And the Spirit of God was hovering over the surface of the waters. 
<em>3</em>And God said, “Let there be light,” and there was light. 
### 2
<em>1</em>Thus the heavens and the earth were completed in all their vast array. 
"
`;